src/language/api/verificationDiagnostics.ts

```diff
@@ -0,0 +1,85 @@
+import { Position } from 'vscode';
+import { Diagnostic, DocumentUri, Range, integer } from 'vscode-languageclient';
+
+export interface IVerificationDiagnosticsParams {
+  uri: DocumentUri;
+  version?: integer;
+  diagnostics: Diagnostic[];
+  perNodeDiagnostic: NodeDiagnostic[];
+  diagnosticsAreResolutionErrors: boolean;
+  perLineDiagnostic: LineVerificationStatus[];
+}
+
+export interface NodeDiagnostic {
+  displayName: string;
+  identifier: string;
+  position: Position;
+  started: boolean;
+  finished: boolean;
+  startTime: integer;
+  endTime: integer;
+  timeSpent: integer;
```

Comment:
In general, this interface has  quite the number of properties. Is it possible to aggregate some of them to at least avoid primitive obsession? I feel that these three values are candidates. If it even was possible to infer `timeSpent` from the other two properties I'd leave it away to reduce the redundancy.

Reply:
TimeSpent is not completely determined by EndTime and StartTime. The formula in the backend is:
```
public int TimeSpent => (int)(Finished ? ((TimeSpan)(EndTime - StartTime)).TotalMilliseconds : Started ? (DateTime.Now - StartTime).TotalMilliseconds : 0);
```
But otherwise, there is no possibilities to add methods to TypeScript interfaces, is it?
Moreover these fields are sent anyway, and I plan to later use them to display something cool in the back-end (like a gant diagram of the proofs)

Reply:
Bear in mind that the interfaces we use here roughly describe the format of a JSON object we receive from the backend and all these type additions are not present at runtime. If we would have something like default-methods, it's probably hard for TypeScript to decide when to use the default method and one from the actual object that was assigned to the variable requiring this interface, to not say impossible. In contrast to other languages, TypeScript does not require an instance to actually implement this interface but only have the shape of it.
Unfortunately, TypeScript does not (yet) have extension methods as we have it in C#.

One last note: That the received JSON object reflects the expected shape described in the interface is nowhere (statically) verified, it could be anything.

Reply:
Indeed ! And based on your comment, I did a slight change now: Instead of saying that the interface returns `Range`, which is problematic because it does not contain methods (it's just a JSON object), I say it returns an `IRange`. I was already doing the conversion for every range, but now it's clearer why it is needed.